lib/IDE/CompletionLookup.cpp

```diff
@@ -1755,6 +1755,42 @@ void CompletionLookup::addNominalTypeRef(const NominalTypeDecl *NTD,
   Builder.setTypeContext(expectedTypeContext, CurrDeclContext);
 }
 
+Type CompletionLookup::getTypeAliasType(const TypeAliasDecl *TAD,
+                                        DynamicLookupInfo dynamicLookupInfo) {
+  // Substitute the base type for a nested typealias if needed.
+  auto ty = getTypeOfMember(TAD, dynamicLookupInfo);
+  auto *typeAliasTy = dyn_cast<TypeAliasType>(ty.getPointer());
+  if (!typeAliasTy)
+    return ty;
+
+
+  // If the underlying type has an error, prefer to print the full typealias,
+  // otherwise get the underlying type.
+  Type underlyingTy = typeAliasTy->getSinglyDesugaredType();
+  if (underlyingTy->hasError())
+    return ty;
+
+  // The underlying type might be unbound for e.g:
+  //
+  // struct S<T> {}
+  // typealias X = S
+  //
+  // Introduce type parameters such that we print the underlying type as
+  // 'S<T>'.
+  return underlyingTy.transformRec([&](TypeBase *ty) -> std::optional<Type> {
+    if (!ty->hasUnboundGenericType()) {
```

Comment:
In fact you don't have to recurse with transformRec() because only the outermost type can be unbound. In this case, the type alias behaves like a generic type alias:
```
typealias X<T> = S<T>
```
Can you handle this like a generic type alias?

Reply:
> In fact you don't have to recurse with transformRec()

Yeah I was just being defensive, I can change to be top-level only

> Can you handle this like a generic type alias?

Would that be any different to `return UGT->getDecl()->getDeclaredInterfaceType()`?

Reply:
Heh actually you could get a non-top-level UnboundGenericType for an underlying type that is also a typealias, maybe we ought to consider recursively looking through to the underlying type though

Reply:
I’m not sure what you mean, like this?

```
typealias A = Array
typealias B = A
```

Reply:
Yeah, we're currently only unwrapping a single level of sugar here so `hasUnboundGenericType` would still be true if we only looked at a top-level UnboundGenericType